File: src/rulesEngine.js
import { locateFieldsOnPage, scanPageForPattern } from './titleBlockLocator.js';
import { escapeRegex } from './util.js';

export function evaluateFieldRules(pages, fieldRules, region) {
  const issues = [];
  for (const p of pages) {
    const fields = locateFieldsOnPage(p, fieldRules, region);
    for (const rule of fieldRules) {
      const result = fields[rule.id];
      if (result.found && (!rule.pattern || result.valid)) continue;

      // The label-based lookup above only finds a field whose exact label
      // text sits inside the configured title-block region right next to
      // the value. Real drawings vary - the title block isn't always in
      // that region, and label wording differs - so before reporting a
      // field as missing or invalid, fall back to scanning the WHOLE page
      // for any text that satisfies the rule's pattern on its own.
      if (rule.pattern && scanPageForPattern(p, rule.pattern) !== null) continue;

      if (!result.found) {
        issues.push({
          category: rule.category, severity: rule.severity, ruleId: rule.id,
          foundText: null, page: p.pageNumber,
          message: `Missing required field "${rule.label}"`,
        });
      } else if (rule.pattern && !result.valid) {
        issues.push({
          category: rule.category, severity: rule.severity, ruleId: rule.id,
          foundText: result.value, page: p.pageNumber,
          message: `Field "${rule.label}" value "${result.value}" does not match expected format`,
        });
      }
    }
  }
  return issues;
}

export function evaluateFormattingRules(pages, formattingRules) {
  const issues = [];
  for (const rule of formattingRules) {
    if (!rule.enabled) continue;
    // `rule.find`/`rule.valid` are assumed to be pre-validated regex strings; a malformed
    // pattern here throws a SyntaxError that propagates out of evaluateRules uncaught. This is
    // deliberately deferred to the future rules-management task that will own regex validation.
    const findRe = new RegExp(rule.find, 'g');
    const validRe = new RegExp(rule.valid);
    for (const p of pages) {
      const text = p.items.map((it) => it.text).join(' ');
      for (const match of text.matchAll(findRe)) {
        if (!validRe.test(match[0])) {
          issues.push({
            category: 'formatting', severity: rule.severity || 'warn', ruleId: rule.id,
            foundText: match[0], page: p.pageNumber, message: rule.message,
          });
        }
      }
    }
  }
  return issues;
}

export function evaluateProjectRules(pages, projectFields, region) {
  const firstPage = pages[0];
  if (!firstPage) return [];
  const requiredFields = projectFields
    .filter((f) => f.value)
    .map((f) => ({ id: f.id, category: 'project', label: f.label, pattern: `^${escapeRegex(f.value)}$` }));
  if (requiredFields.length === 0) return [];

  const fields = locateFieldsOnPage(firstPage, requiredFields, region);
  const issues = [];
  for (const f of requiredFields) {
    const result = fields[f.id];
    if (!result.found || !result.valid) {
      // Same fallback as evaluateFieldRules: the expected text might be on
      // the page but outside the configured region, or not directly next
      // to a recognized label, so check the whole page before giving up.
      if (scanPageForPattern(firstPage, f.pattern) !== null) continue;
      const original = projectFields.find((pf) => pf.id === f.id);
      issues.push({
        category: 'project', severity: 'error', ruleId: f.id,
        foundText: result.value, page: firstPage.pageNumber,
        message: `Project field "${f.label}" expected "${original.value}" but found "${result.value ?? '(missing)'}"`,
      });
    }
  }
  return issues;
}

export function evaluateRules(pages, rulesConfig) {
  const region = rulesConfig.titleBlockRegion;
  const enabledRules = rulesConfig.rules.filter((r) => r.enabled);
  const titleBlockRules = enabledRules.filter((r) => r.category === 'titleBlock');
  const revisionRules = enabledRules.filter((r) => r.category === 'revision');
  const formattingRules = enabledRules.filter((r) => r.category === 'formatting');

  return [
    ...evaluateProjectRules(pages, rulesConfig.project, region),
    ...evaluateFieldRules(pages, titleBlockRules, region),
    ...evaluateFieldRules(pages, revisionRules, region),
    ...evaluateFormattingRules(pages, formattingRules),
  ];
}

File: revit-cad/pyrevit/ARX.extension/lib/arx_rulecore/rules_engine.py
"""Rule evaluation — faithful port of src/rulesEngine.js.

Operates on the abstract page model only, so the exact same logic runs against
PDF (JS), Revit and CAD. Every function returns a list of issue dicts shaped
like the PDF tool's issues.
"""

import re

from .title_block import locate_fields_on_page
from .util import escape_regex


def evaluate_field_rules(pages, field_rules, region):
    issues = []
    for p in pages:
        fields = locate_fields_on_page(p, field_rules, region)
        for rule in field_rules:
            result = fields[rule["id"]]
            if not result["found"]:
                issues.append({
                    "category": rule["category"], "severity": rule["severity"],
                    "ruleId": rule["id"], "foundText": None, "page": p["pageNumber"],
                    "message": 'Missing required field "%s"' % rule["label"],
                })
            elif rule.get("pattern") and not result["valid"]:
                issues.append({
                    "category": rule["category"], "severity": rule["severity"],
                    "ruleId": rule["id"], "foundText": result["value"], "page": p["pageNumber"],
                    "message": 'Field "%s" value "%s" does not match expected format'
                               % (rule["label"], result["value"]),
                })
    return issues


def evaluate_formatting_rules(pages, formatting_rules):
    issues = []
    for rule in formatting_rules:
        if not rule.get("enabled"):
            continue
        find_re = re.compile(rule["find"])
        valid_re = re.compile(rule["valid"])
        for p in pages:
            text = " ".join(it["text"] for it in p["items"])
            for match in find_re.finditer(text):
                hit = match.group(0)
                if not valid_re.search(hit):
                    issues.append({
                        "category": "formatting", "severity": rule.get("severity", "warn"),
                        "ruleId": rule["id"], "foundText": hit, "page": p["pageNumber"],
                        "message": rule["message"],
                    })
    return issues


def evaluate_project_rules(pages, project_fields, region):
    if not pages:
        return []
    first_page = pages[0]
    required_fields = [
        {"id": f["id"], "category": "project", "label": f["label"],
         "pattern": "^%s$" % escape_regex(f["value"])}
        for f in project_fields if f.get("value")
    ]
    if not required_fields:
        return []

    fields = locate_fields_on_page(first_page, required_fields, region)
    issues = []
    for f in required_fields:
        result = fields[f["id"]]
        if not result["found"] or not result["valid"]:
            original = next(pf for pf in project_fields if pf["id"] == f["id"])
            issues.append({
                "category": "project", "severity": "error", "ruleId": f["id"],
                "foundText": result["value"], "page": first_page["pageNumber"],
                "message": 'Project field "%s" expected "%s" but found "%s"'
                           % (f["label"], original["value"], result["value"] or "(missing)"),
            })
    return issues


def evaluate_rules(pages, rules_config):
    region = rules_config["titleBlockRegion"]
    enabled = [r for r in rules_config["rules"] if r.get("enabled")]
    title_block_rules = [r for r in enabled if r["category"] == "titleBlock"]
    revision_rules = [r for r in enabled if r["category"] == "revision"]
    formatting_rules = [r for r in enabled if r["category"] == "formatting"]

    return (
        evaluate_project_rules(pages, rules_config.get("project", []), region)
        + evaluate_field_rules(pages, title_block_rules, region)
        + evaluate_field_rules(pages, revision_rules, region)
        + evaluate_formatting_rules(pages, formatting_rules)
    )
